pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/http/AsyncHttpConnector.java

```
@@ -84,63 +90,106 @@ public boolean keepAlive(Request ahcRequest, HttpRequest request, HttpResponse r
+                        + address.getPort()
+                        + uri.getRawPath();
+        if (uri.getRawQuery() != null) {
+            newUri += "?" + uri.getRawQuery();
+        }
+        return URI.create(newUri);
+    }
 
-            Integer timeout = ClientProperties.getValue(
-                    jerseyRequest.getConfiguration().getProperties(),
-                    ClientProperties.READ_TIMEOUT, 0);
 
-            if (timeout != null && timeout > 0) {
-                resultFuture.get(timeout, TimeUnit.MILLISECONDS);
-            } else {
-                resultFuture.get();
+
+    private boolean resolveRequest(CompletableFuture<ClientResponse> future,
+                                ClientRequest jerseyRequest)
+        throws InterruptedException, ExecutionException, TimeoutException {
+        Future<?> resultFuture = apply(jerseyRequest, new AsyncConnectorCallback() {
+            @Override
+            public void response(ClientResponse response) {
+                future.complete(response);
             }
-        } catch (ExecutionException ex) {
-            Throwable e = ex.getCause() == null ? ex : ex.getCause();
-            throw new ProcessingException(e.getMessage(), e);
-        } catch (Exception ex) {
-            throw new ProcessingException(ex.getMessage(), ex);
+            @Override
+            public void failure(Throwable failure) {
+                future.completeExceptionally(failure);
+            }
+        });
+
+        Integer timeout = ClientProperties.getValue(
+            jerseyRequest.getConfiguration().getProperties(),
+            ClientProperties.READ_TIMEOUT, 0) / 3;
+
+        Object result = null;
+        if (timeout != null && timeout > 0) {
+            result = resultFuture.get(timeout, TimeUnit.MILLISECONDS);
+        } else {
+            result = resultFuture.get();
         }
 
-        return future.join();
+        if (result != null && result instanceof Throwable) {
```

Can you just throw the exception from this method? So that in `apply()` method, you don't need to handle the return result since you already have caught exception and handle the retry logic.

---

Good idea! Thanks